### PR TITLE
Use a jump tables for opcodes

### PR DESCRIPTION
**Pull request by ClementWalter**
## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

We initialize an `instructions: felt*` array with all the opcodes each time we need to run kakarot.

solve #268 

## What is the new behavior?

I've added a (big) jump table in the `decode_and_execute` function. (this is though not really bigger than the previous `generate_instructions` — compared to the previous solution we need a line per index while before it was possible to leave memory holes.

## Other information

Observed a ~15% perf improvement in terms of n_steps

### CODE REVIEW

File: tests/integrations/test_zk_evm.py [hunk 1]
```diff
@@ -61,6 +61,24 @@ async def test_deploy(
         decimals = await erc_20.decimals()
         assert decimals == 18
 
+    async def test_deploy_erc721(
+        self,
+        deploy_solidity_contract: Callable,
+    ):
+        erc_721 = await deploy_solidity_contract(
+            "ERC721", "name", "symbol", caller_address=1
```
Comment (Eikix) [inline tests/integrations/test_zk_evm.py:69]: are these arguments named? 
we could have a more visual setup:
```python
erc_721 = await deploy_solidity_contract(
            contract="ERC721", name="KakarotNft", symbol="KRTNFT", caller_address=1
)
```

Comment (Eikix) [reply tests/integrations/test_zk_evm.py:69]: same for erc20

File: src/kakarot/instructions.cairo [hunk 2]
```diff
@@ -34,80 +34,579 @@ from kakarot.instructions.sha3 import Sha3
...
-    // @param instructions The instruction set.
+    // @dev The function uses an internal jump table to execute the corresponding opcode
     // @param ctx The pointer to the execution context.
     // @return The pointer to the updated execution context.
-
     func decode_and_execute{
         syscall_ptr: felt*,
         pedersen_ptr: HashBuiltin*,
-        range_check_ptr,
+        range_check_ptr: felt,
```
Comment (Eikix) [inline src/kakarot/instructions.cairo:43]: I've never seen this, what does this do?:)

File: src/kakarot/instructions.cairo [hunk 3]
```diff
@@ -34,80 +34,579 @@ from kakarot.instructions.sha3 import Sha3
...
-        let is_pc_le_code_len = is_le_felt(ctx.call_context.bytecode_len, pc);
+        let is_pc_ge_code_len = is_le_felt(ctx.call_context.bytecode_len, pc);
 
-        if (is_pc_le_code_len == 1) {
+        if (is_pc_ge_code_len == TRUE) {
             assert opcode = 0;
         } else {
             assert opcode = [ctx.call_context.bytecode + pc];
         }
+        tempvar offset = 1 + 4 * opcode;
```
Comment (Eikix) [inline src/kakarot/instructions.cairo:59]: could you add a comment to help readability 🪄 ?

File: src/kakarot/instructions.cairo [hunk 4]
```diff
@@ -34,80 +34,579 @@ from kakarot.instructions.sha3 import Sha3
...
+        // Prepare arguments
+        [ap] = syscall_ptr, ap++;
+        [ap] = pedersen_ptr, ap++;
+        [ap] = range_check_ptr, ap++;
+        [ap] = bitwise_ptr, ap++;
+        [ap] = ctx, ap++;
+
+        // call opcode
+        jmp rel offset;
+        call exec_stop;
```
Comment (Eikix) [inline src/kakarot/instructions.cairo]: can you add a comment to state this is opcode number 0? 
when is this called? when offset == 0?

File: src/kakarot/instructions.cairo [hunk 5]
```diff
@@ -34,80 +34,579 @@ from kakarot.instructions.sha3 import Sha3
...
-        }
-
         local opcode;
 
-        let is_pc_le_code_len = is_le_felt(ctx.call_context.bytecode_len, pc);
+        let is_pc_ge_code_len = is_le_felt(ctx.call_context.bytecode_len, pc);
 
-        if (is_pc_le_code_len == 1) {
+        if (is_pc_ge_code_len == TRUE) {
             assert opcode = 0;
```
Comment (Eikix) [inline src/kakarot/instructions.cairo:55]: what happens when opcode <- 0? 
offset <- 1
then jump points to exec_stop ?

**Review by Eikix — CHANGES_REQUESTED** [review]
very cool PR!!
~20% improvement is giga based

File: tests/integrations/test_zk_evm.py (hunk 1 above)
Comment (ClementWalter) [reply tests/integrations/test_zk_evm.py:69]: yes they could be kwargs as well, they are directly given to the `web3.contract().constructor()`.
I can update this, but tbh I'd like to move ERC20 and ERC721 tests into their own files integrating all the tests from solmate repo, so I'm note sure it's worth doing anything here.

But it makes me realized that these utils aren't well documented, so in any case I'm adding some docstring, and consequently why not these args as kwargs

File: src/kakarot/instructions.cairo (hunk 2 above)
Comment (ClementWalter) [reply src/kakarot/instructions.cairo:43]: it's the same, I think that I changed this for an easier copy/pasting, which led to the useless cast at line 590 😇

File: src/kakarot/instructions.cairo [hunk 6]
```diff
@@ -34,80 +34,579 @@ from kakarot.instructions.sha3 import Sha3
...
-
-        // Revert if pc < 0
-        with_attr error_message("Kakarot: InvalidCodeOffset") {
-            assert_nn(pc);
-        }
-
         local opcode;
 
-        let is_pc_le_code_len = is_le_felt(ctx.call_context.bytecode_len, pc);
+        let is_pc_ge_code_len = is_le_felt(ctx.call_context.bytecode_len, pc);
```
Comment (spapinistarkware) [inline src/kakarot/instructions.cairo:52]: is_felt_le is not that cheap. Use the 128bit version if possible.

Comment (ClementWalter) [reply src/kakarot/instructions.cairo:52]: good to know, thx

**Review by Eikix — APPROVED** [review]
lgtm

**Pull request merged**